src/xlsx.js

```diff
@@ -0,0 +1,130 @@
+      _: {value: workbook},
+      sheetNames: {
+        value: workbook.worksheets.map((sheet) => sheet.name),
+        writable: false,
+        enumerable: true,
+      },
+    });
+  }
+  sheet(name, {range, headers = false} = {}) {
+    const sname =
+      typeof name === "number"
+        ? this.sheetNames[name]
+        : this.sheetNames.includes(name + "")
+        ? name + ""
+        : null;
+    if (sname == null) throw new Error(`Sheet not found: ${name}`);
+    const sheet = this._.getWorksheet(sname);
+    return extract(sheet, {range, headers});
+  }
+}
+
+function extract(sheet, {range, headers}) {
+  let [[c0, r0], [c1, r1]] = parseRange(range, sheet);
+  const headerRow = headers && sheet._rows[r0++];
+  let names = new Set(["#"]);
+  for (let n = c0; n <= c1; n++) {
+    let name = (headerRow ? valueOf(headerRow._cells[n]) : null) || AA(n);
+    while (names.has(name)) name += "_";
+    names.add(name);
+  }
+  names = new Array(c0).concat(Array.from(names));
+
+  const output = new Array(r1 - r0 + 1);
+  for (let r = r0; r <= r1; r++) {
+    const row = (output[r - r0] = Object.defineProperty({}, "#", {
+      value: r + 1,
+    }));
+    const _row = sheet._rows[r];
+    if (_row && _row.hasValues)
+      for (let c = c0; c <= c1; c++) {
+        const value = valueOf(_row._cells[c]);
+        if (value != null) row[names[c + 1]] = value;
+      }
+  }
+
+  output.columns = names.filter(() => true); // Filter sparse columns
+  return output;
+}
+
+function valueOf(cell) {
+  if (!cell) return;
+  const {value} = cell;
+  if (value && value instanceof Date) return value;
+  if (value && typeof value === "object") {
+    if (value.formula || value.sharedFormula)
+      return value.result && value.result.error ? NaN : value.result;
+    if (value.richText) return value.richText.map((d) => d.text).join("");
+    if (value.text)
+      return value.hyperlink
+        ? `<a href="${encodeURI(value.hyperlink)}">${escapeHTML(
+            value.text
+          )}</a>`
+        : value.text;
+    return value;
+  }
+  return value;
+}
+
+function escapeHTML(string) {
+  return string.replace(/[&<>"']/g, (m) => {
+    switch (m) {
+      case "&":
+        return "&amp;";
+      case "<":
+        return "&lt;";
+      case ">":
+        return "&gt;";
+      case '"':
+        return "&quot;";
+      case "'":
+        return "&#39;";
+    }
+  });
+}
+
+function parseRange(specifier = [], {columnCount, rowCount}) {
+  if (typeof specifier === "string") {
+    if (!specifier.includes(":")) throw new Error("Malformed range specifier");
+    const [[c0 = 0, r0 = 0], [c1 = columnCount - 1, r1 = rowCount - 1]] =
+      specifier.split(":").map(NN);
+    return [
+      [c0, r0],
+      [c1, r1],
+    ];
+  } else if (typeof specifier === "object") {
+    const [
+      [c0 = 0, r0 = 0] = [],
+      [c1 = columnCount - 1, r1 = rowCount - 1] = [],
+    ] = specifier;
+    return [
+      [c0, r0],
+      [c1, r1],
+    ];
```

Comment:
Do we need to support an object form as well as a string form for _specifier_? Is the string form (e.g., `B4:AF234`) the more familiar of the two? I’d guess so, so I’d favor just having the string form here unless there’s a strong precedent.

In addition to being simpler, here are two more reasons to use only a string. First, it’s a simpler structure that is easier to validate. For example, I can crash the code with `TypeError: null is not an object` by saying `parseRange([null, null], {})` or `parseRange(null, {})` and it’d be better to have a nice error message. Second, it lets us coerce the specifier to a string so that objects that implement toString can be used as range specifiers if desired. This is a nitch feature to be sure, but idiomatic JavaScript given how nearly everything is typeof object. (Plot has a more specific test for object literals [using toString](https://github.com/observablehq/plot/blob/8b47987bbc52ec08d897c3bf39af26a1b5835d85/src/mark.js#L186-L192). You could also use Array.isArray here on the specifier as a more specific check, but my point is it’s better to avoid type checking at all if possible.)

Reply:
Yeah, I think you're right and we should drop the array form. Pretty sure the more useful, efficient, and friendly form is the string one. The array one I think I only included for the spreadsheet-averse people.

Reply:
The array form can be useful for someone who would create a UI (say, with sliders or a brush) to specify the range. (But we could say they'd just have to copy the AA function).